### PR TITLE
Fix self assign copy move

### LINKED ISSUE #54: Fix all self assign/copy/moves.
In many case there are custom copy and move assignment and constructors that are not save for assignment to self. Lets fix these.

Copy / Move Constructor
-------------------------
In normal situations self-construction does not happen without invoking undefined behaviour.
Lets add an axiom at the start of the constructor to check for self-construction.

```
tt_axiom(&other != this)
```

Copy / Move Assignment
-------------------------
In normal code it is possible for self-assignment. even the standard algorithms may use std::swap with both arguments referencing the same object.

The best way to solve this is to make sure that self-assignment simply works with the normal code, without any check. In this case we need an internal comment to mark the fact that we checked this:

```
// Self-assignment is allowed.
```

Otherwise a check is needed, which returns *this when self-assignment is happening, we mark this fact specifically 'unlikely' because it simply shouldn't really be happening:

```
tt_short_circuit_self_assignment(other);
```

The `tt_short_circuit_self_assingment()` should probably be a macro in required.hpp:
```
#define tt_short_circuit_self_assignment(other)\
  if (&(other) == this) [[unlikely]] return *this;
```

### PR DESCRIPTION
**Pull request by jhalakpatel**
Fixes #54

### CODE REVIEW

**Comment by PinkFlufflyLlama** [comment]
I really don't think that your macro adds any more clarity to the code. In fact I would go as far to say that it makes the code more unclear as at first glance you may not see the macro and thusly not expect it to return early at that point.

File: src/ttauri/pixel_map.hpp [hunk 1]
```diff
@@ -211,6 +211,7 @@ class pixel_map {
 
     pixel_map &operator=(pixel_map &&other) noexcept
     {
+        // Self assignment is allowed.
```
Comment (takev) [inline src/ttauri/pixel_map.hpp:214]: Remove the comment below it. Which meant the same

**Review by takev — CHANGES_REQUESTED** [review]

File: src/ttauri/required.hpp [hunk 2]
```diff
@@ -42,4 +42,7 @@ constexpr ssize_t operator"" _z(unsigned long long lhs) noexcept
     return static_cast<ssize_t>(lhs);
 }
 
+#define tt_short_circuit_self_assignment(other)\
```
Comment (takev) [inline src/ttauri/required.hpp:45]: After talking with @PinkFlufflyLlama we think it is better if we call this `tt_return_on_self_assignment`. If you would like to change this that would be great.

If not, I will still merge this and change it at a later time.

File: src/ttauri/pixel_map.hpp (hunk 1 above)
Comment (jhalakpatel) [reply src/ttauri/pixel_map.hpp:214]: Addressed in d2100bc455c4e29e86b1a744defb8229b1de469e

File: src/ttauri/required.hpp (hunk 2 above)
Comment (jhalakpatel) [reply src/ttauri/required.hpp]: Addressed in cb9d45a540e34629d3b276cc91fc50f2097c3e5d.

**Review by takev — APPROVED** [review]

**Pull request merged**